engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertAllNode.java

```diff
@@ -0,0 +1,73 @@
+package org.enso.interpreter.runtime.data.hash;
+
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import org.enso.common.LanguageInfo;
+import org.enso.interpreter.node.expression.builtin.meta.EqualsNode;
+import org.enso.interpreter.node.expression.builtin.meta.HashCodeNode;
+
+@GenerateUncached
+@NodeInfo(
+    shortName = "MapInsertAll",
+    description = "Inserts all elements from the given container into a hash map",
+    language = LanguageInfo.ID)
+public abstract class HashMapInsertAllNode extends Node {
+
+  public static HashMapInsertAllNode build() {
+    return HashMapInsertAllNodeGen.create();
+  }
+
+  public static HashMapInsertAllNode getUncached() {
+    return HashMapInsertAllNodeGen.getUncached();
+  }
+
+  /**
+   * Insert all the elements from the given container into the given map.
+   *
+   * @param self A map to insert into.
+   * @param container Either a map, or a list of pairs to insert into the map.
+   * @param maxItems Maximum number of items to insert into the map from the container.
+   */
+  public abstract EnsoHashMap executeInsertAll(
+      VirtualFrame frame, EnsoHashMap self, EnsoHashMap container, int maxItems);
+
+  @Specialization
+  EnsoHashMap doEnsoHashMaps(
+      VirtualFrame frame,
+      EnsoHashMap self,
+      EnsoHashMap other,
+      int maxItems,
+      @Cached HashCodeNode hashCodeNode,
+      @Cached EqualsNode equalsNode) {
+    assert maxItems > 0;
+    var selfSize = self.getHashSize();
+    var otherSize = other.getHashSize();
+    if (otherSize == 0) {
+      return self;
+    }
+    var mapBuilder = EnsoHashMapBuilder.createWithCapacity(selfSize + otherSize);
+
+    var selfMapBuilder = self.getMapBuilder(frame, true, hashCodeNode, equalsNode);
+    var selfEntriesIt = selfMapBuilder.getEntriesIterator(selfMapBuilder.generation());
```

Comment:
Using `selfMapBuilder.generation()` isn't correct. Fixed in https://github.com/enso-org/enso/pull/12201/files#r1946171034